watershed.py

```diff
@@ -0,0 +1,334 @@
+    while not heap.is_empty:
+        elem = heap.pop()
+        if compact:
+            if output[elem.index] and elem.index != elem.source:
+                # non-marker, already visited, move on to next item
+                continue
+            output[elem.index] = output[elem.source]
+        for i in range(n_neighbors):
+            # get the flattened address of the neighbor
+            if affinities:
+                # offsets are 2d (size, 2) with columns 0 and 1 corresponding to 
+                # affinities and image neighbour indices respectively
+                neighbor_index = offsets[i, 1] + elem.index
+                # in this case the index used to find elem.value will be 2d tuple
+                affinity_index = tuple(offsets[i] + np.array([0, elem.index]))
+            else:
+                neighbor_index = offsets[i] + elem.index
+            if not mask[neighbor_index]:
+                # neighbor is not in mask, move on to next neighbor
+                continue
+            if output[neighbor_index]:
+                # if there is a non-zero value in output, move on to next neighbor
+                continue
+            # if the neighbor is in the mask and not already labeled, add to queue
+            age += 1
+            new_elem = Element()
+            if affinities:
+                new_elem.value = image_raveled[affinity_index]
+            else:
+                new_elem.value = image_raveled[neighbor_index]
+            if compact:
+                new_elem.value += (compactness *
+                                       _euclid_dist(neighbor_index, elem.source,
+                                                        strides))
+            output[neighbor_index] = output[elem.index]
+            new_elem.age = age
+            new_elem.index = neighbor_index
+            new_elem.source = elem.source
+            heap.push(new_elem)
+    return output
+
+
+def _euclid_dist(pt0, pt1, strides):
+    result, curr = 0, 0
+    for i in range(strides.shape[0]):
+        curr = (pt0 // strides[i]) - (pt1 // strides[i])
+        result += curr * curr
+        pt0 = pt0 % strides[i]
+        pt1 = pt1 % strides[i]
+    return np.sqrt(result)
+    
+
+class Element:
+    def __init__(self, value=None, index=None, age=None, source=None):
+        self._value = value
+        self._index = index
+        self._age = age
+        self._source = source
+    @property
+    def value(self):
+        return self._value
+    @value.setter
+    def value(self, v):
+        self._value = v
+    @property
+    def index(self):
+        return self._index
+    @index.setter
+    def index(self, i):
+        self._index = i
+    @property
+    def age(self):
+        return self._age
+    @age.setter
+    def age(self, a):
+        self._age = a
+    @property
+    def source(self):
+        return self._source
+    @source.setter
+    def source(self, s):
+        self._source = s
+
+
+class Heap:
```

Comment:
Whoa! This is simultaneously cute and terrible :joy: 

It's awesome that you implemented a heap from scratch, but it's theoretically terrible (not your fault!) to sort on each push. Actually looking at the code you haven't really implemented a heap but something else entirely, since your data are sorted. Heaps don't have to be sorted, just partially sorted. See https://en.wikipedia.org/wiki/Heap_(data_structure)

I don't know how much you know about algorithmic complexity, but sorting is O(n log(n)), so if you do it for all n pixels, you get a cost of O(n^2 logn(n)).

In contrast, adding an item to a heap is O(log(n)), so doing it for all n pixels is (n log(n)). So that's much better.

Now, O() hides away the constant factors (how much does one pixel actually cost), which are probably much lower for NumPy arrays compared to Python. But, we should still do it the Right Way and then improve the constant factors using Cython & co. The nice thing is that there is a heap implementation in the Python standard library, so you can use that directly!

https://docs.python.org/3.7/library/heapq.html